usb/ehci/ehci.c

```diff
@@ -0,0 +1,737 @@
+			while (node->period * 2 < pipe->interval)
+				node->period *= 2;
+		}
+		fprintf(stderr, "ehci: allocQh period: %d\n", node->period);
+	}
+
+	if (pipe->type == usb_transfer_control && pipe->dev->speed != usb_high_speed)
+		qh->ctrlEp = 1;
+
+	node->last = &qh->transferOverlay;
+	node->qh = qh;
+
+	return qh;
+}
+
+
+static void ehci_allocPeriodicBandwidth(ehci_t *ehci, struct qh_node *node)
+{
+	struct qh_node *tmp;
+	unsigned int i, n, best;
+	unsigned int ucnt[8] = { 0 };
+
+	best = (unsigned)-1;
+	node->phase = 0;
+	node->uframe = 0xff;
+
+	/* Find the best periodicList index (phase) to begin Qh linking */
+	for (i = 0; i < node->period && i < EHCI_PERIODIC_SIZE; i++) {
+		n = 0;
+		/* Count Qhs linked to this periodic index */
+		tmp = ehci->periodicNodes[i];
+		while (tmp != NULL) {
+			n++;
+			tmp = tmp->next;
+		}
+
+		if (n < best) {
+			best = n;
+			node->phase = i;
+		}
+	}
+	fprintf(stderr, "ehci: allocPeriodicBandwidth period: %d phase: %d\n", node->period, node->phase);
+
+	/* Find the best microframe in a frame. For periods equal to 1, send it every microframe */
+	if (node->qh->epSpeed == usb_high_speed && node->period > 1) {
+		tmp = ehci->periodicNodes[node->phase];
+		while (tmp != NULL) {
+			if (tmp->uframe != 0xff)
+				ucnt[tmp->uframe]++;
+			tmp = tmp->next;
+		}
+
+		best = (unsigned)-1;
+		node->uframe = 0;
+		for (i = 0; i < 8; i++) {
+			if (ucnt[i] < best) {
+				node->uframe = i;
+				best = ucnt[i];
+			}
+		}
+	}
+}
+
+
+static void ehci_linkPeriodicQh(hcd_t *hcd, struct qh_node *node)
+{
+	ehci_t *ehci = (ehci_t *)hcd->priv;
+	struct qh_node *t;
+	int i;
+
+	mutexLock(ehci->periodicLock);
+	ehci_allocPeriodicBandwidth(ehci, node);
+	node->qh->smask = (node->uframe != 0xff) ? (1 << node->uframe) : 0xff;
+	/* TODO: Handle SPLIT transactions */
+	node->qh->cmask = 0xff;
+
+	t = ehci->periodicNodes[node->phase];
+	while (t && t->next != NULL && t->next->period >= node->period)
+		t = t->next;
+
+	if (t == NULL || t->period < node->period) {
+		/* New first element */
+		node->next = ehci->periodicNodes[node->phase];
+
+		for (i = node->phase; i < EHCI_PERIODIC_SIZE; i += node->period) {
+			ehci->periodicNodes[i] = node;
+			ehci->periodicList[i].pointer = va2pa(node->qh) >> 5;
+			ehci->periodicList[i].type = ehci_item_qh;
+			ehci->periodicList[i].terminate = 0;
+		}
+	}
+	else {
+		/* Insert inside */
+		node->next = t->next;
+		t->next = node;
+		node->qh->horizontal.pointer = va2pa(node->next->qh) >> 5;
+		node->qh->horizontal.type = ehci_item_qh;
+		node->qh->horizontal.terminate = 0;
+
+		t->qh->horizontal.pointer = va2pa(node->qh) >> 5;
+		t->qh->horizontal.type = ehci_item_qh;
+		t->qh->horizontal.terminate = 0;
+	}
+
+	if (node->next == NULL) {
+		/* New last element */
+		node->qh->horizontal.terminate = 1;
+	}
+
+	mutexUnlock(ehci->periodicLock);
+}
+
+
+static void ehci_linkAsyncQh(hcd_t *hcd, struct qh_node *node)
+{
+	ehci_t *ehci = (ehci_t *)hcd->priv;
+	int first = ehci->asyncList == NULL;
```

Comment:
Not good for readability. Perhaps:
```suggestion
	int first = (ehci->asyncList == NULL) ? 0 : 1;
```